ports/espressif/common-hal/analogbufio/BufferedIn.c

```diff
@@ -0,0 +1,283 @@
+#define ADC_READ_TIMEOUT_MS         2000
+
+#if defined(CONFIG_IDF_TARGET_ESP32)
+#define ADC_RESULT_BYTE     2
+#define ADC_CONV_LIMIT_EN   1   //For ESP32, this should always be set to 1
+#elif defined(CONFIG_IDF_TARGET_ESP32S2)
+#define ADC_RESULT_BYTE     2
+#define ADC_CONV_LIMIT_EN   0
+#elif defined(CONFIG_IDF_TARGET_ESP32C3) || defined(CONFIG_IDF_TARGET_ESP32H2)
+#define ADC_RESULT_BYTE     4
+#define ADC_CONV_LIMIT_EN   0
+#elif defined(CONFIG_IDF_TARGET_ESP32S3)
+#define ADC_RESULT_BYTE     4
+#define ADC_CONV_LIMIT_EN   0
+#endif
+
+static adc_digi_convert_mode_t convert_mode = ADC_CONV_SINGLE_UNIT_2;
+static adc_digi_output_format_t output_format = ADC_DIGI_OUTPUT_FORMAT_TYPE1;
+static uint8_t adc_channel = 0;
+
+void common_hal_analogbufio_bufferedin_construct(analogbufio_bufferedin_obj_t *self, const mcu_pin_obj_t *pin, uint32_t sample_rate) {
+    uint16_t adc1_chan_mask = 0;
+    uint16_t adc2_chan_mask = 0;
+
+    if( self->pin != NULL) {
+        mp_raise_ValueError(translate("ADC DMA already initialized"));
+    }
+
+    output_format = ADC_DIGI_OUTPUT_FORMAT_TYPE1;
+    if(pin->adc_index == ADC_UNIT_1) {
+        convert_mode = ADC_CONV_SINGLE_UNIT_1;
+    } else {
+        convert_mode = ADC_CONV_SINGLE_UNIT_2;
+    }
+
+    if (pin->adc_index == NO_ADC || pin->adc_channel == NO_ADC_CHANNEL) {
+        raise_ValueError_invalid_pin();
+    }
+
+    adc_channel = pin->adc_channel;
+
+    /*
+     *	Chip version	Conversion Mode		Output Format Type
+     *	ESP32			1					TYPE1
+     *	ESP32S2			1,2,BOTH,ALTER		TYPE1, TYPE2
+     *	ESP32C3			ALTER				TYPE2
+     *	ESP32S3			1,2,BOTH,ALTER		TYPE2
+     *	ESP32H3			1,2,BOTH,ALTER		TYPE2
+     */
+
+#if defined(CONFIG_IDF_TARGET_ESP32)
+    if(pin->adc_index != ADC_UNIT_1) {
+        mp_raise_ValueError(translate("ESP32 only supports ADC1 unit"));
+    }
+#endif
+#if defined(CONFIG_IDF_TARGET_ESP32S2)
+#endif
+#if defined(CONFIG_IDF_TARGET_ESP32C3)
+    //ESP32C3 only supports alter mode
+    convert_mode = ADC_CONV_ALTER_UNIT;
+    output_format = ADC_DIGI_OUTPUT_FORMAT_TYPE2;
+#endif
+#if defined(CONFIG_IDF_TARGET_ESP32S3)
+#endif
+#if defined(CONFIG_IDF_TARGET_ESP32H2)
+#endif
+
+    self->pin = pin;
+    common_hal_mcu_pin_claim(pin);
+
+    if(pin->adc_index == ADC_UNIT_1) {
+        adc1_chan_mask = 1 << pin->adc_channel;
+    } else {
+        adc2_chan_mask = 1 << pin->adc_channel;
+    }
+
+    adc_digi_init_config_t adc_dma_config = {
+        .max_store_buf_size = DMA_BUFFER_SIZE,
+        .conv_num_each_intr = NUM_SAMPLES_PER_INTERRUPT,
+        .adc1_chan_mask = adc1_chan_mask,
+        .adc2_chan_mask = adc2_chan_mask,
+    };
+
+#if defined(DEBUG_ANALOGBUFIO)
+    mp_printf(&mp_plat_print,"pin:%d, ADC channel:%d, ADC index:%d, adc1_chan_mask:0x%x, adc2_chan_mask:0x%x\n",pin->number,pin->adc_channel,pin->adc_index,adc1_chan_mask,adc2_chan_mask);
+#endif //DEBUG_ANALOGBUFIO
+    esp_err_t err = adc_digi_initialize(&adc_dma_config);
```

Comment:
I think you'll want to do all of the below initialization at read time. Doing it here will prevent a second bufferedin on a second pin. if you do all the init and deinit in the read, then you an share amongst multiple objects.

Reply:
Thanks your for your suggestion! The intention of this PR is to support recording audio via ADC DMA https://github.com/adafruit/circuitpython/issues/7454

The python application would typically look as follows:
Initialize the analogbufio
Read and process samples in a loop
Dinitialize analogbufio

Loop in second step above would be required on ESP32 boards without PSRAM support where amount of RAM is limited. The ADC DMA samples would be read into same buffer again and again and would be processed in smaller chunks.  If the init and deinit is done in read itself then such a loop to collect samples continuously won't be possible.

Reply:
@tannewt may comment further next week on this. I'm not sure the other `analogbufio` has this use case. So we may want to make the semantics consistent, perhaps even through an API change or addition. For instance, maybe ` continuing` or `continuous` or some similarly-named boolean flag might make sense.

Reply:
If supporting DMA on multiple pins simultaneously is the original use case then we need to brainstorm on how to handle this in case of ESP32. The ESP32 DMA engine provides use of two masks for ADC1 and ADC2 units which selects pins to be used for DMA operation. Based on pins selected by this mask the DMA starts collecting samples from these pins. The DMA read result provided by `adc_digi_read_bytes` consists of bits for DMA Unit, channel number (pin) and sample value. So the application has to separate the data for the pins based on values provided by the result. Since current analogbufio only supports configuration of one pin at a time, I have  implemented the module with support for one pin a time as there is no way to specify the mask to select multiple pins. Further we need to address how to segregate the results from `adc_digi_read_bytes` and buffer them for individual pins for respective analogbufio instances. It should also be noted that the valid pins that could be configured for DMA differs across the ESP32 chip version.

Reply:
Hello @tannewt awaiting further comments from you.  Can we document the restriction of using one pin at a time for ESP32 port of analogbufio for now and move ahead?

Reply:
Sorry for the slow response. It was a three day weekend here in the US and I didn't have the brain power yesterday to context switch to this.

Unfortunately, I don't think the existing API is good enough for audio input. I'd suggest following my original suggestion to do init/deinit for reading here. Then, follow up with a new `AudioIn` API. It'll need to take a buffer in the constructor, return samples read without blocking and detect overflows. It'll be pretty different than this API.

Reply:
Thanks for your response! I hope you enjoyed your long weekend. Coming back to thread, even if I follow your original suggestion of init/deinit we wont' be able to support second bufferedIn on second pin because mask for ADC units will get overwritten and previous bufferedIn samples would be stopped if second bufferedIn is configured. Even if we remember the mask (read/modify/write) and configure second BufferedIn appropriately, the samples collected via adc_digi_read_bytes would be interleaved for both pins and we can't separate them for two different bufferedIn objects. This is because the adc_digi_read_bytes done by bufferedIn read operation for first pin would throw away samples collected for second pin. Thus the init/deinit change does not allow for second bufferedIn as you originally suggested. Let me know if you need further explanation on this.

Reply:
Yup! It was good. The read should occur during `readinto` and that ensures only one pin is being read at a time. This API isn't designed for background reading.

Reply:
Okay. I get it. The current analogbufio APIs are designed for "one shot" or "single conversion" mode of ADC DMA. It does not support "continuous conversion" mode. I will make changes as per your suggestion to move init/deinit in readinto.

Reply:
@tannewt  I have addressed your review comment of initializing and de-initializing dma in readinto function via commits https://github.com/adafruit/circuitpython/pull/7602/commits/09f84e351318f9f3bf5bff6bb9664703ca743c78 and https://github.com/adafruit/circuitpython/pull/7602/commits/9e73e8351aace95c6600345ca80134a7e13915de

Reply:
Thank you!